### PR TITLE
Adding vgg16 model built using VGG-Blocks

### PR DESCRIPTION
**Pull request by adithya-tp**
Hey there! I wanted to get familiar with this repository, and adding I felt adding vgg16 would be a good step in this direction. The code and folder structure for this pull request follows that of #50 . The only new thing I've contributed are the "VGG blocks" function in VGG16.hpp , that makes building the VGG-16 net easier in VGG16_impl.hpp. 

From what I've read in the literature, one of the significant contributions from the vgg paper was using such repeatable blocks to build the net easily. I feel like this is not addressed in #28 which adds vgg19. However, I do love the tutorial from that pull request. I'd love to volunteer coding up vgg19 using the vgg blocks as well.

My pull request is still a work in progress. I just wanted to put this out there, so that people don't waste time duplicating work. 

I understand this repo is currently undergoing restructuring, and I'd be happy to refactor my code according to the folder structure that we end up deciding on.

I still need to test the network. What is stopping me right now is I don't know how to get cmake to add the executable to the existing "build/bin" folder (I'm just recently getting familiar with cmake 😅).
Feel free to drop in your suggestions. I'm looking forward to incorporating your thoughts and ideas to get this thing working.

### CODE REVIEW

**Comment by adithya-tp** [comment]
Okay I've understood how to create the executable using cmake. I am in the process of testing out the vgg16 network.

**Comment by kartikdutt18** [comment]
I think network should work irrespective of input size, what do you think?
Thanks.

**Comment by adithya-tp** [comment]
> I think network should work irrespective of input size, what do you think?
> Thanks.

Yeah i guess so. I got the network to compile and run. However, the loss is quite high - so if anyone comes across anything fishy in the code, do share it here. Currently the training is very slow (i don't know if its because vgg is running on cpu)

**Comment by kartikdutt18** [comment]
Things that I would try:
1. Better optimisation policy.
2. Different step sizes (something I did to make LeNet converge).
3. Better Initialization of weights (rather than Random).
You could try different permutations that would make it converge.
I made a makeshift DarkNet19 that plan to open a PR for, so if that can converge (kind of), I think this can too. You can try various permutations and eventually find the best parameters.
Hope this helps.

**Comment by adithya-tp** [comment]
> Things that I would try:
> 
> 1. Better optimisation policy.
> 2. Different step sizes (something I did to make LeNet converge).
> 3. Better Initialization of weights (rather than Random).
>    You could try different permutations that would make it converge.
>    I made a makeshift DarkNet19 that plan to open a PR for, so if that can converge (kind of), I think this can too. You can try various permutations and eventually find the best parameters.
>    Hope this helps.

Thanks for the suggestions! I've got a few queries about your darknet model.
1. What was your initial loss like? (my initial loss was around 10^7)
2. How long did an epoch take approximately?
3. What was your lowest loss? (I'm assuming you used mnist)

Also if you could find some logical flaw in my code, that'd be quite helpful as well. Thanks!

**Comment by kartikdutt18** [comment]
With wrong params I got an initial loss of 10^8, when I experimented a bit I got to initial loss at 10 and lowest 0.07. I am using my college workstation which is amazingly fast so it 400s. Locally it takes about 10 minutes or so (not sure). And yes, I used mnist.
I will go through the code tonight or tomorrow morning.

**Comment by adithya-tp** [comment]
> With wrong params I got an initial loss of 10^8, when I experimented a bit I got to initial loss at 10 and lowest 0.07. I am using my college workstation which is amazingly fast so it took 400s. Locally it takes about 10 minutes or so (not sure). And yes, I used mnist.
> I will go through the code tonight or tomorrow morning.

Oh nice! Sorry to bother you over this, but could you share the hyperparameters you used? So that I can have a point of reference. Thanks.

**Comment by kartikdutt18** [comment]
> Update (after commit "Changed to adam optimizer and tweaked hyperparameters"): Well I tweaked the optimization policy and hyperparameters. Now the loss starts at around 1.3 rather than the previous 10^17. However it still takes a looot of time to train. Am I still missing something? Or is it just that VGG is a decent sized network, that its slow to train on my machine? I'll try to delve into this further.
Great. Did you try training with OpenBLAS / NVBLAS?

**Comment by adithya-tp** [comment]
> Great. Did you try training with OpenBLAS / NVBLAS?

No I haven't. I wasn't aware such a thing existed. Could you point me towards some code snippets that use this? That'd be really helpful to get an idea

**Comment by kartikdutt18** [comment]
Hey @adithya-tp, you can find the executable in bin folder with cmake that is there. All you need to do is link it BLAS after you compile armadillo with BLAS. I  used a single file so I don't have any cmake for it. I think this should be [helpful](http://arma.sourceforge.net/faq.html).

**Comment by favre49** [comment]
Models are now to be moved to mlpack/models, and this has become an examples folder. I think this would better be suited there. Sorry for the inconvenience, but please migrate your PR to that repository.

**Comment by adithya-tp** [comment]
> Models are now to be moved to mlpack/models, and this has become an examples folder. I think this would better be suited there. Sorry for the inconvenience, but please migrate your PR to that repository.

Have been quite occupied with assignments and quizzes. Will migrate this PR soon! Thanks for the heads up :+1:

**Comment by adithya-tp** [comment]
Closing this, since I've migrated it to [#5](https://github.com/mlpack/models/pull/5) in the models repo.
